### PR TITLE
Allow installation on PHP 8.

### LINKED ISSUE #54: PHP 8.0 support
### Feature Request

|    Q        |   A
|------------ | ------
| New Feature | yes

#### Summary

To be prepared for the december release of PHP 8.0, this repository has some additional TODOs to be tested against the new major version.

In order to make this repository compatible, one has to follow these steps:

- [ ] Modify `composer.json` to provide support for PHP 8.0 by adding the constraint `~8.0.0`
- [ ] Modify `composer.json` to drop support for PHP less than 7.3
- [ ] Modify `composer.json` to implement phpunit 9.3 which supports PHP 7.3+
- [ ] Modify `.travis.yml` to ignore platform requirements when installing composer dependencies (simply add `--ignore-platform-reqs` to `COMPOSER_ARGS` env variable)
- [ ] Modify `.travis.yml` to add PHP 8.0 to the matrix (*NOTE*: Do not allow failures as PHP 8.0 has a feature freeze since 2020-08-04!)
- [ ] Modify source code in case there are incompatibilities with PHP 8.0

### PR DESCRIPTION
**Pull request by ADmad**
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | yes
| QA            | no

### CODE REVIEW

File: composer.json
Comment (Ocramius) [inline composer.json]: Should probably first upgrade the test suite

File: .travis.yml
Comment (Ocramius) [inline .travis.yml]: If you added `^8.0`, then doing `--ignore-platform-reqs` here shouldn't be required

**Review by Ocramius — CHANGES_REQUESTED** [review]

File: .travis.yml
Comment (ADmad) [reply .travis.yml]: It's required temporarily as other dev deps like `php-http/psr7-integration-tests` and `laminas/laminas-zendframework-bridge` don't get installed without it as their deps also need to be updated to allow installing on PHP 8.

https://travis-ci.com/github/laminas/laminas-diactoros/jobs/372193687#L225

File: composer.json
Comment (ADmad) [reply composer.json]: Currently diactoros is PHP 7.1+ while phpunit 9.3 is PHP 7.3+, so I was trying to see if both phpunit versions can be supported without bumping up diactoros' PHP version constraint :slightly_smiling_face:.

Comment (Ocramius) [reply composer.json]: Yeah, we can bump to `^7.3` in the next minor

Comment (ADmad) [reply composer.json]: Currently CakePHP 4 supports PHP 7.2+ and has diactoros as dependency. So unless really necessary we would prefer if diactoros does not goto 7.3+ just due to phpunit :)

Supporting phpunit 8.5 and 9.3 together with php 7.2+ isn't too difficult, we have done it https://github.com/cakephp/cakephp/pull/14885.

Comment (Ocramius) [reply composer.json]: Cake can support an older version of diactoros, and then you'll be fine 👍

Comment (ADmad) [reply composer.json]: Only if the older version supports PHP 8 too.

Comment (Ocramius) [reply composer.json]: Why would it need to? If you pin to `^2.3.1`, you get `7.1.*` support, as well as PHP 8 support (which will happen with `2.4.0`, according to this patch)

**Comment by ADmad** [comment]
Related PRs for dependencies:

https://github.com/laminas/laminas-zendframework-bridge/pull/69
https://github.com/php-http/psr7-integration-tests/pull/45
https://github.com/http-interop/http-factory-tests/pull/44

**Comment by ADmad** [comment]
@Ocramius I have bumped PHP constraint to 7.3+ as per your feedback. Temporarily using my fork to run the test suite until `php-http/psr7-integration-tests` has a new release.

The two test failures aren't directly related to the PHP version bump and probably need to be addressed separately. 

The docblock of [StreamFactoryInterface::createStreamFromFile()](https://github.com/php-fig/http-factory/blob/1a2099a23a7bca67f13c0634723137a364a3ebf8/src/StreamFactoryInterface.php#L30) says `@throws \RuntimeException If the file cannot be opened.` and the test case correctly tests for that. But `Laminas\Diactoros\StreamFactory::createStreamFromFile()` always throws an instance of `Laminas\Diactoros\Exception\InvalidArgumentException`  (which extends `\InvalidArgumentException` if it fails to create a stream.

**Comment by ADmad** [comment]
Regarding the failing tests this commit message has a quoted response for the PHP-FIG member: https://github.com/http-interop/http-factory-tests/pull/42/commits/c29f9c5a6b0c26c2aff25cbe46502aa9eeaf8157.

It's seems throwing a `RuntimeException` instead of `InvalidArgumentException` exception when failing to create a stream is recommend. So maybe the exception thrown by diactoros could be changed for next minor release?

Another option would be to see if http-interop is willing to just widen the asserted exception to `Exception` instead of `RuntimeException` as done for other tests in the commit linked above.

**Comment by ADmad** [comment]
Can someone get the maintainers of `php-http/psr7-integration-tests` to merge or at least respond to https://github.com/php-http/psr7-integration-tests/pull/45 so that we can make progress on this PR?

**Comment by weierophinney** [comment]
> Can someone get the maintainers of `php-http/psr7-integration-tests` to merge

I've bumped via a comment; crossing fingers for now.

**Comment by weierophinney** [comment]
> Regarding the failing tests this commit message has a quoted response for the PHP-FIG member: [http-interop/http-factory-tests@c29f9c5](https://github.com/http-interop/http-factory-tests/commit/c29f9c5a6b0c26c2aff25cbe46502aa9eeaf8157).
> 
> It's seems throwing a `RuntimeException` instead of `InvalidArgumentException` exception when failing to create a stream is recommend. So maybe the exception thrown by diactoros could be changed for next minor release?
> 
> Another option would be to see if http-interop is willing to just widen the asserted exception to `Exception` instead of `RuntimeException` as done for other tests in the commit linked above.

Third option: we override that test case in our own test suite, to address the behavior.

**Comment by ADmad** [comment]
> Third option: we override that test case in our own test suite, to address the behavior.

Not quite sure how to do that since `http-factory-tests`'s test suite is directly run https://github.com/laminas/laminas-diactoros/blob/e609ac206649cc51f00e9728314067848706dc33/phpunit.xml.dist#L7

**Comment by ADmad** [comment]
BTW should this PR now target `2.5.x` branch? Given the min. PHP version bump I believe a new minor release would be warranted.

**Comment by ADmad** [comment]
> BTW should this PR now target 2.5.x branch?

I took the liberty of doing so since the milestone for the PR was already changed to `2.5.0`.

**Comment by ADmad** [comment]
I don't know your backward compatibility policy is but since these changes are going into a new minor release perhaps we can just update `Stream::setStream()` to throw `RuntimeException` instead of `InvalidArgument` exception to fix the testsuite failure.

**Comment by ADmad** [comment]
I updated to use the new `php-http/psr7-integration-tests` release. Tweaked `Stream::isValidStreamResourceType()` to account for GD changes in PHP 8.

**Comment by andypost** [comment]
Handy guide about changes https://php.watch/versions/8.0

**Comment by ADmad** [comment]
>https://github.com/http-interop/http-factory-tests/commit/c29f9c5a6b0c26c2aff25cbe46502aa9eeaf8157
>
> It's seems throwing a RuntimeException instead of InvalidArgumentException exception when failing to create a stream is recommend. So maybe the exception thrown by diactoros could be changed for next minor release?

My recommendation would be to do the above to fix the test failures. I can make required changes if I get confirmation.

**Comment by andypost** [comment]
Would be great to get in next minor release, upcoming Drupal 9.1 using it as dev-dependency, so would be great to declare full compatibility

**Comment by froschdesign** [comment]
@andypost 
> Would be great to get in next minor release, upcoming Drupal 9.1 using it as dev-dependency, so would be great to declare full compatibility

Everyone is welcome to help!

**Comment by andypost** [comment]
Checked 2 remaining failures - both caused by https://php.watch/versions/8.0/internal-function-exceptions

I see 2 ways to solve it

1. add condition to `\Laminas\Diactoros\Stream::setStream` before `fopen()` https://github.com/laminas/laminas-diactoros/blob/2.5.x/src/Stream.php#L345 - Drupal used to fix this way to keep backward compatibility https://www.drupal.org/project/drupal/issues/3177541

2. change expectations in https://github.com/http-interop/http-factory-tests/pull/45
- `\Interop\Http\Factory\StreamFactoryTestCase::testCreateStreamFromNonExistingFile()`
- `\Interop\Http\Factory\StreamFactoryTestCase::testCreateStreamFromInvalidFileName`

**Comment by andypost** [comment]
Also filed issue for warnings https://github.com/php-http/psr7-integration-tests/pull/47

<details>
<summary>It allowed to pass 1 test</summary>

```
# composer test
> phpunit --colors=always
PHPUnit 9.4.2 by Sebastian Bergmann and contributors.

Warning:       Your XML configuration validates against a deprecated schema.
Suggestion:    Migrate your XML configuration using "--migrate-configuration"!

.............................................................   61 / 1101 (  5%)
.............................................................  122 / 1101 ( 11%)
.............................................................  183 / 1101 ( 16%)
.............................................................  244 / 1101 ( 22%)
.............................................................  305 / 1101 ( 27%)
.............................................................  366 / 1101 ( 33%)
.............................................................  427 / 1101 ( 38%)
.............................................................  488 / 1101 ( 44%)
.............................................................  549 / 1101 ( 49%)
.............................................................  610 / 1101 ( 55%)
.............................................................  671 / 1101 ( 60%)
.............................................................  732 / 1101 ( 66%)
.............................................................  793 / 1101 ( 72%)
.............................................................  854 / 1101 ( 77%)
.............................................................  915 / 1101 ( 83%)
.............................................................  976 / 1101 ( 88%)
............................................................. 1037 / 1101 ( 94%)
.....................................................F....... 1098 / 1101 ( 99%)
...                                                           1101 / 1101 (100%)

Time: 00:01.745, Memory: 12.00 MB

There was 1 failure:

1) Interop\Http\Factory\StreamFactoryTest::testCreateStreamFromNonExistingFile
Failed asserting that exception of type "Laminas\Diactoros\Exception\InvalidArgumentException" matches expected exception "PHPUnit\Framework\Error\Warning". Message was: "Invalid stream reference provided" at
/srv/src/Stream.php:350
/srv/src/Stream.php:65
/srv/src/StreamFactory.php:41
/srv/vendor/http-interop/http-factory-tests/test/StreamFactoryTestCase.php:95
.

FAILURES!
Tests: 1101, Assertions: 1834, Failures: 1.
Script phpunit --colors=always handling the test event returned with error code 1

```

</details>

**Comment by andypost** [comment]
The last failed test is because test expects `RuntimeException` but `InvalidArgumentException` extends `LogicException` which is not runtime

**Comment by andypost** [comment]
Following hack for  `http-interop/http-factory-tests` allowed to pass tests but looks instead of it the `Exception\InvalidArgumentException` should extend `\RuntimeException`

```diff
diff --git a/test/StreamFactoryTestCase.php b/test/StreamFactoryTestCase.php
index 4641788..5c9578a 100644
--- a/test/StreamFactoryTestCase.php
+++ b/test/StreamFactoryTestCase.php
@@ -86,7 +86,7 @@ abstract class StreamFactoryTestCase extends TestCase
         $filename = $this->createTemporaryFile();
         unlink($filename);
 
-        $this->expectException(RuntimeException::class);
+        $this->expectException(Exception::class);
         $stream = $this->factory->createStreamFromFile($filename);
     }
 
```

**Comment by ADmad** [comment]
> Everyone is welcome to help!

@froschdesign I had been awaiting the Laminas team's response regarding how to deal with the `RuntimeException` vs `InvalidArgumentException` issue. With none forth coming I finally just went ahead and updated the code to throw `RuntimeException`.

**Comment by andypost** [comment]
Both PRs for dependencies are accepted but no releases yet

File: composer.json
Comment (boesing) [inline composer.json]: Please use `~8.0.0` as we internally decided to go that way for PHP 8.0.

File: src/Stream.php
Comment (boesing) [inline src/Stream.php:350]: As this is a BC break, I would prefer having some kind of "feature toggle" which allows upstream consumers to "enable" the `RuntimeException`

We could mark that feature toggle method as deprecated aswell as stating that with the next major, we will completely switch to `RuntimeException`.

Thoughts @weierophinney?

File: src/Stream.php
Comment (boesing) [inline src/Stream.php:377]: As this is the only change in here, I have a few questions:

- does all of these functions work on that `GdImage` object?
  - `fclose`
  - `stream_get_contents`
  - `fread`
  - `stream_get_meta_data`

If the answer for any of these functions is "no", we need to find a completely new way to make the `Stream`-implementation compatible with the new `GdImage` object as we are using `fclose` in `Stream::close`, `stream_get_contents` in `Stream::getContents` and `stream_get_metadata` in some of the other methods...

**Review by boesing — CHANGES_REQUESTED** [review]

File: composer.json
Comment (alexpott) [reply composer.json]: @boesing is there a record of the discussion about using ~8.0.0? That means when 8.1 comes out any of your dependencies will require a new release of their laminas dependency for them to be installable even if there are no code changes required to the laminas library. This feels a step backwards. I think the approach recently taken by Symfony to change their constraints to `>=` is the way to go but I can understand libraries choosing to go with `^8.0` - but ~8.0.0 feels very restrictive.

Comment (Ocramius) [reply composer.json]: >  That means when 8.1 comes out any of your dependencies will require a new release of their laminas dependency for them to be installable even if there are no code changes required to the laminas library.

This is intentional: PHP is too unstable to pin to a major release, and indeed, restricting is intentional. As for symfony, I can only say that jumping off the bridge because somebody told you to do it is not a wise approach to decision-making.

If all is good, next year, it is just a matter of bumping dependency ranges: until then, there is no guarantee that 8.1 will be compatible :-)

As for the meeting where this was decided, I think it was this one: https://github.com/laminas/technical-steering-committee/blob/main/meetings/minutes/2020-09-07-TSC-Minutes.md#hacktoberfest

Can't find the precise point in any meeting notes though - would need to check the `#tsc-meeting` channel on https://laminas.slack.com/

File: src/Stream.php
Comment (ADmad) [reply src/Stream.php:377]: I checked locally on PHP 7.3. The following test cases

```
    public function testIsReableWithGdResource()
    {
        $resource = imagecreate(1, 1);
        $stream   = new Stream($resource);
        $this->assertTrue($stream->isReadable());
    }

    public function testCloseWithGdResource()
    {
        $resource = imagecreate(1, 1);
        $stream   = new Stream($resource);
        $this->assertNull($stream->close());
    }

```
result in 
```
1) LaminasTest\Diactoros\StreamTest::testIsReableWithGdResource
stream_get_meta_data(): supplied resource is not a valid stream resource

2) LaminasTest\Diactoros\StreamTest::testCloseWithGdResource
fclose(): supplied resource is not a valid stream resource
```

So the existing code is already broken and doesn't work with image resources.

Comment (ADmad) [reply src/Stream.php:377]: So if you do want to support using image resources it should be sorted out separately as fixing existing broken functionality is outside the scope of this PR.

File: composer.json
Comment (alexpott) [reply composer.json]: @Ocramius I'm not jumping off a bridge because somebody told me too. I'm being stopped sending out scouts across a bridge because someone is unnecessarily putting a stop sign in the way.

There are a number of reasons why I think constraining the maximum version of PHP a library can run on is a not healthy for the ecosystem. Here are a some:
* Increases friction to early adopters from testing your code and creating PRs to fix it
* Increases friction to dependencies testing against new PHP versions. They have to jump through hoops and if nothing is broken they then require a release from Laminas in order to run tests. It makes the ecosystem form a long chain and, if you are on the end of the chain, you have a lot waiting to do.
* Something hand wavy about software freedoms, gophp5, and Postel's law. It just feels wrong.
* Composer's PHP constraints don't really work this way anyway... if there's something like `"platform": { "php": 7.4.99 }` in the root composer.json then, even though I'm running PHP 8, composer will still get laminas dependencies and not inform me a project is incompatible.

I can understand limiting to major versions ie something like `^7.0` (although I disagree with it) but `~8.0.0` increases the friction for laminas's dependencies due to a vague "PHP is too unstable to pin to a major release". I've worked on making Drupal 7.0, 7.1, 7.2, 7.4 and 8.0 compatible. It's a very large codebase and has many dependencies. PHP is generally the least of the problems. The biggest problem in maintaining compatibility against multiple PHP versions at the moment is PHPUnit.

For example, any dependency of laminas-feed that wants to be compatible with PHP 8.0 is currently waiting on https://github.com/laminas/laminas-feed/pull/28 - which is only about code quality - and the PHP 8.0 compatibility PR only stopped PHP from issuing a deprecation message (a good thing to be sure) BUT now in order to be PHP 8 compatible your dependencies have to upgrade from 2.12.x to 2.13.0 - when there is no new functionality. There's a similar story for laminas-escaper and laminas-diactoros.

Comment (ADmad) [reply composer.json]: > The biggest problem in maintaining compatibility against multiple PHP versions at the moment is PHPUnit.

This! For e.g. the only reason PHP version contraint for this package needed to be updated to `^7.3` is because of PHPUnit.

I don't know in what utopia Sebastian lives where the whole PHP ecosystem gets updated as soon as active support for a PHP version ends.

File: src/Stream.php
Comment (boesing) [reply src/Stream.php:377]: You are right, I just wanted to verify that we have this covered.
I will try to keep in mind that I create an Issue regarding this.

**Review by boesing — APPROVED** [review]

File: composer.json
Comment (boesing) [reply composer.json]: No one stops you from using composer v2 with `--ignore-platform-req=php`.
So what is the real pain you want to express?

We just want to specify every PHP version as tested. This can only work if we have active tests against that version. So every user of laminas can be 100% sure (as of PHP 8.0) that the code is tested under the current version the user executes `composer update/install`.

Imho, thats a huge promise to all laminas users and gives a maximum of "safety". Bugs can rise even with tested versions but at least it is compatible and no-one has to check if a vendor package of laminas/mezzio/laminas-api-tool is compatible with PHP 8.
Users are not beta-testers for new PHP versions but actually thats what symfony and most other libs are doing. They let their users beta-test a probably unstable version of their package. Imho, thats the point what we should talk about and not about "friction".

**And this is what I call a healthy PHP ecosystem**

Comment (froschdesign) [reply composer.json]: @alexpott
> For example, any dependency of laminas-feed that wants to be compatible with PHP 8.0 is currently waiting on laminas/laminas-feed#28 - which is only about code quality

Unfortunately this is not entirely correct. This pull request is completed but [the commits needs a correct signed off](https://github.com/laminas/laminas-feed/pull/28/checks?check_run_id=1336868041) by the creator of the pull request.
But I hope that we can solve the problem on laminas-feed quickly.

Comment (alexpott) [reply composer.json]: I've just found https://github.com/php-http/psr7-integration-tests/pull/45 and I'm more than a bit bemused. But to find the laminas project asking for PHP 8 compatibility on project because it is holding back the ecosystem is a bit much. The points made are exactly the points I'm making about needing compatibility before a release in order to be able to test before. I quote...
> bump
>
> RC2 just dropped, so PHP 8 is reaching relative stability. A number of PSR-7 providers are trying to ensure they are ready (e.g., Diactoros, Slim), but we are all currently blocked until this is merged.

Diactoros has been sitting on approved merge requests for days, and similarly for Feeds (who's PHP 8 compatibility release is blocked on a new integration with Psalm...). This means that unless Drupal makes exceptions to its release management policies it'll be June 2021 before we have an officially PHP 8 compatible release.

Also given the time its taken to get 8.0.0 compatibility done setting your constraint to ~8.0.0 does not fill your dependencies with confidence that you'll 8.1.0 compatible releases in place in time for your dependencies to be 8.1.0 compatible.

**Comment by dereuromark** [comment]
Projects and libraries need this PR to get merged in order to avoid any --ignore-platform-req hacks when finalizing PHP 8 compatibility.
Is there anything else needed to get this done at this point?

File: src/Stream.php
Comment (weierophinney) [reply src/Stream.php:350]: I'd consider this a bugfix, but call it out in the "Changed" section of the changelog, and note the _why_ of the change ("to comply with PSR-7 integration test suites")

File: composer.json
Comment (weierophinney) [reply composer.json]: These are all valid view points. However, there's an aspect that those _consuming_ the library are missing from those _maintaining_ the library: Using `^` constraints makes maintaining the library far harder.

How?

Because every time we get a new minor or major PHP version, we have to increase our test matrix. And every time we increase the test matrix, we generally have to revisit the PHPUnit versions we use and the syntax each uses to ensure that tests will actually run. This makes even simple contributions often more complex because new failures arise that were not present before; additionally, if the contributor does not have every version of PHP currently supported by the library on hand, they may not even know there are test failures until CI/CD runs... which, when the matrix gets large, take longer and longer to complete. And we all know that when we have to wait for results, we tend to get bored and move on.

Yes, in an **ideal** world, there would be no BC breaks or deprecations within a given minor release series, whether that's PHP or libraries you use. But in point of fact, these arise **all the time**. That's the origin of the "PHP is too unstable to pin to a major release'" statement. Every single minor release that comes out, we MUST do a cycle to test our libraries against it, because inevitably some new deprecation or change **does** lead to breakage, whether that's simply a test failing because of a new deprecation notice, or, worse, **users** identifying issues. When a user jumps to a new version and suddenly the library does not work, that immediately becomes **our** problem, but now there's urgency to it.

On top of that, there's the question of how long we support a given minor or major release of a library. If the library has versions 1.0, 1.1, 1.2, 2.0, 2.1, 3.0, and 3.1, and all of them have a minimum supported version of `^7.0`, does that mean we have to provide bugfixes and security fixes to _all_ of these branches? or just the latest minor releases of each major branch (e.g., 1.2, 2.1, and 3.1)? or current major and previous major?

By narrowing the window of PHP versions we support (via the `~` constraint, or bumping the minimum minor we support), we make the maintenance story far easier:

- We provide security patches to branches where the minimum supported PHP version is still supported by community PHP.
- We can have a smaller number of PHPUnit versions we have to be compatible with.
- There's a smaller test matrix we need to fulfill.
- When a new minor or major version of PHP is released, we create a new minor release that changes the PHP constraints, to signal that we have actually _tested_ it on that version, versus _assumed_ it would remain compatible.

**Comment by weierophinney** [comment]
> Projects and libraries need this PR to get merged in order to avoid any --ignore-platform-req hacks when finalizing PHP 8 compatibility.
> Is there anything else needed to get this done at this point?

Reviewing today, and will likely get a release out.

**Review by weierophinney — APPROVED** [review]
:ship:

File: composer.json
Comment (alexpott) [reply composer.json]: > we generally have to revisit the PHPUnit versions we use and the syntax each uses to ensure that tests will actually run.

Fortunately the world has changed... https://twitter.com/s_bergmann/status/1328197689029386240

Comment (ADmad) [reply composer.json]: Finally, Hallelujah!

**Comment by GrahamCampbell** [comment]
The upstream change is wrong. You need to catch the ValueError from fopen on PHP 8.

**Comment by froschdesign** [comment]
@GrahamCampbell 
Please open a new issue report because this pull request is already closed. Thanks in advance! 👍

**Comment by GrahamCampbell** [comment]
A fix will be made to the upstream package, which will break this package's tests when it is tagged.

**Pull request merged**
